### ISSUE TITLE
Feature switch with ILLink substitution does not always trim branches

### ISSUE THREAD

**Issue by eiriktsarpalis** [dotnet/runtime#90034]
@eerhardt @vitek-karas for context to https://github.com/dotnet/runtime/pull/90013/files#r1284314974 adding this method call resulted in rooting the reflection resolver.

_Originally posted by @eiriktsarpalis in https://github.com/dotnet/runtime/pull/90013#discussion_r1284316652_

**Comment by eiriktsarpalis** [comment]
> And this is all just for feature switches which are pretty rare (I know it doesn't seem like it, because in the BCL we use them quite often, but we're the exception here).

That's true, however `JsonSerializer.IsReflectionEnabledByDefault` is public API that we're explicitly documenting as being substituted by the linker. The idea is that library authors (including aspnetcore already) rely on this flag to avoid accidentally rooting the substantial STJ reflection components.

Perhaps the solution is to be more deliberate in documenting what types of branching patterns are "safe" in that regard.

**Comment by vitek-karas** [comment]
> Perhaps the solution is to be more deliberate in documenting what types of branching patterns are "safe" in that regard.

Exactly - we just need to define that set first. So far it's been a "best effort" type of solution... obviously now we need something more defined.
